### PR TITLE
SwiftCompilerSources: support for ForwardingInstructions, ConversionInstructions, OwnershipTransitionInstruction

### PR DESCRIPTION
**Pull request by atrick**
No functional change intended. But this does fix holes in the existing utilities and make the Swift/C++ code base more consistent. That might improve the power of some analyses which could affect SIL patterns.

Namely, AccessUtils/MemAccessUtils now finds the actual reference root consistently.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

File: SwiftCompilerSources/Sources/SIL/ForwardingInstruction.swift [hunk 1]
```diff
@@ -397,3 +397,33 @@ extension LinearFunctionExtractInst: ForwardingInstruction {
...
   public var canForwardOwnedValues: Bool { true }
 }
+
+// -----------------------------------------------------------------------------
+// ownership transition instructions
+
+/// An instruction that transfers lifetime dependence from a single operand to a single result. The operand and result
+/// have the same identity, but they are not part of the same forwarded lifetime:
+/// copy_value, move_value, begin_borrow.
+public protocol OwnershipTransitionInstruction: UnaryInstruction {
```
Comment (meg-gupta) [inline SwiftCompilerSources/Sources/SIL/ForwardingInstruction.swift:407]: nit: These are in "ForwardindInstruction".swift. Is there any other more appropriate file to put them ?

File: lib/SIL/Utils/SILBridging.cpp [hunk 2]
```diff
@@ -402,11 +402,6 @@ static_assert((int)BridgedMemoryBehavior::MayWrite == (int)swift::MemoryBehavior
 static_assert((int)BridgedMemoryBehavior::MayReadWrite == (int)swift::MemoryBehavior::MayReadWrite);
 static_assert((int)BridgedMemoryBehavior::MayHaveSideEffects == (int)swift::MemoryBehavior::MayHaveSideEffects);
 
-static_assert((int)BridgedInstruction::AccessKind::Init == (int)swift::SILAccessKind::Init);
```
Comment (meg-gupta) [inline lib/SIL/Utils/SILBridging.cpp:405]: nit: can we keep the static_assert that c++ enum values are same as swift ones ?

**Review by meg-gupta — APPROVED** [review]
LGTM

File: SwiftCompilerSources/Sources/SIL/ForwardingInstruction.swift (hunk 1 above)
Comment (atrick) [reply SwiftCompilerSources/Sources/SIL/ForwardingInstruction.swift:407]: No strong opinion. But I think it's important to understand the relationship between ForwardingInstruction and OwnershipTransitionInstruction. It's easy to get them confused and an instruction is either one or the other but never both

File: lib/SIL/Utils/SILBridging.cpp (hunk 2 above)
Comment (atrick) [reply lib/SIL/Utils/SILBridging.cpp:405]: > nit: can we keep the static_assert that c++ enum values are same as swift ones ?

That would be great. Can you tell me how to do it?

**Comment by atrick** [comment]
The deserializer crashed. Something to do with the swift version. Trying again.

**Comment by atrick** [comment]
@swift-ci test

File: lib/SIL/Utils/SILBridging.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/SIL/Utils/SILBridging.cpp:405]: I think it should work if you keep the original `AccessKind` in `SILBridging.h` and then create the swift enum this way: 

```
public var accessKind: AccessKind {
   AccessKind(rawValue: bridged.BeginAccessInst_getAccessKind().rawValue)!
 }
```

File: lib/SILOptimizer/PassManager/PassPipeline.cpp [hunk 3]
```diff
@@ -61,18 +61,22 @@ static llvm::cl::opt<bool> SILViewSILGenCFG(
     "sil-view-silgen-cfg", llvm::cl::init(false),
     llvm::cl::desc("Enable the sil cfg viewer pass before diagnostics"));
 
-llvm::cl::opt<bool> EnableDeinitDevirtualizer(
-    "enable-deinit-devirtualizer", llvm::cl::init(false),
-    llvm::cl::desc("Enable the DeinitDevirtualizer pass."));
+static llvm::cl::opt<bool>
+    EnableDestroyHoisting("enable-destroy-hoisting", llvm::cl::init(false),
```
Comment (eeckstein) [inline lib/SILOptimizer/PassManager/PassPipeline.cpp:65]: This looks like a wrong rebase: the EnableDestroyHoisting option was recently removed by Nate

File: SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LetPropertyLowering.swift [hunk 4]
```diff
@@ -202,10 +202,28 @@ private extension RefElementAddrInst {
 }
 
 private extension Value {
+  func isReferenceDerived(from root: Value) -> Bool {
```
Comment (eeckstein) [inline SwiftCompilerSources/Sources/Optimizer/FunctionPasses/LetPropertyLowering.swift:205]: I think `referenceIsDerived(from:)` would be a better name

**Review by eeckstein — COMMENTED** [review]
lgtm

File: lib/SIL/Utils/SILBridging.cpp (hunk 2 above)
Comment (meg-gupta) [reply lib/SIL/Utils/SILBridging.cpp:405]: Huh, I guess that's not sufficient either to truly ensure these enum values  will be equivalent.

Comment (meg-gupta) [reply lib/SIL/Utils/SILBridging.cpp:405]: And way too complex and costly. So please ignore my comment!

Comment (eeckstein) [reply lib/SIL/Utils/SILBridging.cpp:405]: Maybe add a comment for `SILAccessKind` and `BeginAccessInst.AccessKind` that both must be kept in sync

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
@swift-ci clean smoke test

**Comment by eeckstein** [comment]
I'll take a look

**Comment by atrick** [comment]
> I'll take a look

The bootstrapping1 compiler is miscompiling SIL.o.
The miscompiled function is FunctionEffects.description. This works around the problem:
```
+++ b/SwiftCompilerSources/Sources/SIL/Effects.swift
@@ -31,6 +31,7 @@ public struct FunctionEffects : CustomStringConvertible, NoReflectionChildren {
     self.sideEffects = nil
   }

+  @_optimize(none)
   public var description: String {
     var numArgEffects = escapeEffects.arguments.reduce(0, { max($0, $1.argumentIndex) }) + 1
     if let sideEffects = sideEffects {
```

**Comment by atrick** [comment]
The call to `sideEffects.arguments.count` is being miscompiled.

I narrowed the problem down to a couple lines of code in a separate function that should be easy to compare SIL output and bisect:

```
+++ b/SwiftCompilerSources/Sources/SIL/Effects.swift
@@ -31,7 +31,11 @@ public struct FunctionEffects : CustomStringConvertible, NoReflectionChildren {
     self.sideEffects = nil
   }

+  @_optimize(none)
   public var description: String {
+    // print("REAL SIDE EFFECT COUNT \(sideEffects == nil ? 0 : sideEffects!.arguments.count)")
+    return getArgEffectsDescription() ? "GOOD EFFECT\n" : "BAD EFFECT\n"
+    /*
     var numArgEffects = escapeEffects.arguments.reduce(0, { max($0, $1.argumentIndex) }) + 1
     if let sideEffects = sideEffects {
       numArgEffects = max(numArgEffects, sideEffects.arguments.count)
@@ -56,6 +60,20 @@ public struct FunctionEffects : CustomStringConvertible, NoReflectionChildren {
       result += "[global: \(sideEffects.global)]\n"
     }
     return result
+     */
+  }
+
+  //!!!
+  @_silgen_name("__effects_decription")
+  @inline(never)
+  func getArgEffectsDescription() -> Bool {
+    var numArgEffects = escapeEffects.arguments.reduce(0, { max($0, $1.argumentIndex) }) + 1
+    //!!! After running the reduce loop above, sideEffects!.arguments.count is now magically 0!
+    // print("SIDEEFFECTCOUNT \(sideEffects == nil ? -1 : sideEffects!.arguments.count)")
+    if let sideEffects = sideEffects {
+      return sideEffects.arguments.count > 0
+    }
+    return false
   }
 }
```

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
Failed Tests (1):
  Swift-validation(linux-x86_64) :: stdlib/RangeSet.swift

# *** Signal 6: Backtracing from 0x7f6a88ffc00b... done ***

**Comment by atrick** [comment]
@swift-ci test linux platform

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Review by atrick — COMMENTED** [review]
/cc @gottesmm 
[Fix MarkUnresolvedReferenceBindingInst handling](https://github.com/apple/swift/pull/71826/commits/a048455ad4d77fc484bbf723053d08f0360ff38f)

**Pull request merged**